Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.h

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2022 Apple Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <WebCore/PixelBuffer.h>
+
+namespace WebKit {
+
+class SharedMemory;
+
+class ShareablePixelBuffer : public WebCore::PixelBuffer {
+public:
+    static RefPtr<ShareablePixelBuffer> tryCreate(const WebCore::PixelBufferFormat&, const WebCore::IntSize&);
+
+    SharedMemory& data() const { return m_data.get(); }
+    Ref<SharedMemory>&& takeData() { return WTFMove(m_data); }
```

Comment:
should this be a std::exchange?

Reply:
I realize this function is not used. So I deleted it. But I found a similar function in ByteArrayPixelBuffer. I think the prototype of the other function is not right. ByteArrayPixelBuffer::takeData() is returning the Ref<JSC::Uint8ClampedArray> by-move. This will reset the pointer of the m_data to nullptr. So ByteArrayPixelBuffer itself will be valid but it m_data is in weird state. Its pointer is nullptr so it won't release anything. But it is still accessible. So I changed it to be like this:

`static Ref<JSC::Uint8ClampedArray> takeData(Ref<ByteArrayPixelBuffer>&& pixelBuffer) { return WTFMove(pixelBuffer->m_data); }`

By doing this, we take the Uint8ClampedArray out of the ByteArrayPixelBuffer before releasing it. And after this call, pixelBuffer will be invalid Ref and can't be used.

This function will call Ref(Ref&& other) which will call Ref::leak(). Ref::leak will call `PtrTraits::exchange(m_ptr, nullptr)`.

I think I can't call std::exchange for m_data since it is Ref<JSC::Uint8ClampedArray>. I can't exchange it with a nullptr.